stage1/init/init.go

```diff
@@ -295,6 +307,165 @@ func forwardedPorts(pod *Pod) ([]networking.ForwardedPort, error) {
+	f, err := os.Open("/proc/cgroups")
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	sc := bufio.NewScanner(f)
+
+	sc.Scan()
+
+	cgroups := make(map[int][]string)
+	for sc.Scan() {
+		var controller string
+		var hierarchy int
+		var num int
+		var enabled int
+		fmt.Sscanf(sc.Text(), "%s %d %d %d", &controller, &hierarchy, &num, &enabled)
+
+		if _, ok := cgroups[hierarchy]; !ok {
+			cgroups[hierarchy] = []string{controller}
+		} else {
+			cgroups[hierarchy] = append(cgroups[hierarchy], controller)
+		}
+	}
+
+	if err := sc.Err(); err != nil {
+		return nil, err
+	}
+
+	return cgroups, nil
+}
+
+func getControllers(cgroups map[int][]string) []string {
+	var controllers []string
+	for _, cs := range cgroups {
+		controllers = append(controllers, strings.Join(cs, ","))
+	}
+
+	return controllers
+}
+
+func getControllerSymlinks(cgroups map[int][]string) map[string]string {
+	symlinks := make(map[string]string)
+
+	for _, cs := range cgroups {
+		if len(cs) > 1 {
+			tgt := strings.Join(cs, ",")
+			for _, ln := range cs {
+				symlinks[ln] = tgt
+			}
+		}
+	}
+
+	return symlinks
+}
+
+func createCgroups(root string, machineID string, appHashes []types.Hash) error {
+	cgroups, err := parseCgroups()
+	if err != nil {
+		return fmt.Errorf("error parsing /proc/cgroups: %v", err)
+	}
+
+	controllers := getControllers(cgroups)
+
+	// 1. Mount /sys read-only
+	sys := filepath.Join(root, "/sys")
+	if err := os.MkdirAll(sys, 0700); err != nil {
+		return err
+	}
+	if err := syscall.Mount("sysfs", sys, "sysfs", syscall.MS_RDONLY|syscall.MS_NOSUID|syscall.MS_NOEXEC|syscall.MS_NODEV, ""); err != nil {
+		return fmt.Errorf("error mounting %q: %v", sys, err)
+	}
+
+	// 2. Mount /sys/fs/cgroup
+	cgroupTmpfs := filepath.Join(root, "/sys/fs/cgroup")
+	if err := os.MkdirAll(cgroupTmpfs, 0700); err != nil {
+		return err
+	}
+	if err := syscall.Mount("tmpfs", cgroupTmpfs, "tmpfs", syscall.MS_NOSUID|syscall.MS_NOEXEC|syscall.MS_NODEV|syscall.MS_STRICTATIME, "mode=755"); err != nil {
+		return fmt.Errorf("error mounting %q: %v", cgroupTmpfs, err)
+	}
+
+	// 4. Mount controllers
+	for _, c := range controllers {
+		// 4a. Mount controller
+		cPath := filepath.Join(root, "/sys/fs/cgroup", c)
+		if err := os.MkdirAll(cPath, 0700); err != nil {
+			return err
+		}
+		if err := syscall.Mount("cgroup", cPath, "cgroup", syscall.MS_NOSUID|syscall.MS_NOEXEC|syscall.MS_NODEV, c); err != nil {
+			return fmt.Errorf("error mounting %q: %v", cPath, err)
+		}
+
+		// 4b. Check if we're running from a unit to know which subcgroup
+		// directories to mount read-write
+		var subcgroup string
+		fromUnit, err := runningFromUnitFile()
+		if err != nil {
+			return fmt.Errorf("error determining if we're running from a unit file: %v", err)
+		}
+		if fromUnit {
+			unit, err := getUnitFileName()
+			if err != nil {
+				return fmt.Errorf("error getting unit name: %v", err)
+			}
+			subcgroup = filepath.Join(cPath, "system.slice", unit, "system.slice")
```

Comment:
isn't this counter to https://github.com/coreos/rkt/issues/911#issuecomment-103853978 ?